ramips: fix kernel oops in `mt7621_nfc_write_page_hwecc`

anton-ryzhov opened the pull request:
https://bugs.openwrt.org/index.php?do=details&task_id=3416

`mt7621_nfc_write_page_hwecc` may be called with `buf=NULL`, but `mt7621_nfc_check_empty_page` always tries to read it
That caused Oops `Unable to handle kernel paging request at virtual address 00000000`.

I'm not sure about potential corner cases of that solution, but that worked for me, marked block successfully.

adschm commented:
Hi, please also add a header with description for the patch (file) itself, like done for many of the other patches in the same dir.

Please also add a `Fixes: FS#3416` in the commit message directly before the Signed-off-by line.

adschm commented:
Apart from that, this looks to me like something that should be fixed directly in kernel and then backported?

ptpt52 commented:
This code looks very confusing
ping @dengqf6

981213 commented:
@adschm mt7621_nand isn't an upstream driver. It's rejected there because upstream guys want the support for mt7621 added into existing mtk_nand driver instead.
We took it because we were using another really messy driver from the old MTK SDK, and this one is better than that.

@anton-ryzhov Instead of adding a patch to fix an existing patch, could you change the original patch directly instead?

adschm commented:
Sorry, wasn't looking close enough initially; in this case, I also think we should edit 0300-mtd-rawnand-add-driver-support-for-MT7621-nand-flash.patch directly.

OT: If it's rejected anyway, we should consider moving drivers/mtd/nand/raw/mt7621_nand.c from patches to files ...

anton-ryzhov committed:
ramips: fix kernel oops in `mt7621_nfc_write_page_hwecc`

`mt7621_nfc_write_page_hwecc` may be called with `buf=NULL`, but `mt7621_nfc_check_empty_page` always tries to read it
That caused Oops `Unable to handle kernel paging request at virtual address 00000000`

Fixes: FS#3416

Signed-off-by: Anton Ryzhov <anton@ryzhov.me>

anton-ryzhov commented:
@adschm 

> Please also add a Fixes: FS#3416 in the commit message directly before the Signed-off-by line.

>  in this case, I also think we should edit 0300-mtd-rawnand-add-driver-support-for-MT7621-nand-flash.patch directly.

Done

@ptpt52 

> This code looks very confusing

What do you mean?

981213 commented:
@anton-ryzhov 
Hi!
Got some feedbacks from the driver author @hackpascal :
mt7621_nfc_check_empty_page should be patched to skip data checking for a null buf argument instead.
It's supposed to check if upper layer is trying to write nothing into a page. It should still check oob when kernel don't want to write any data.

anton-ryzhov commented:
@981213 Did I get you right? https://github.com/openwrt/openwrt/pull/3552/files/b081927e269d713de9e8f7f8ece3f5286848dd6a

981213 commented on target/linux/ramips/patches-5.4/0300-mtd-rawnand-add-driver-support-for-MT7621-nand-flash.patch:
```diff
@@ -1155,9 +1155,11 @@ Signed-off-by: Weijie Gao <weijie.gao@mediatek.com>
++		for (i = 0; i < mtd->writesize; i++)
++			if (buf[i] != 0xff)
++				return 0;
++	}
```
you've changed the  number of lines without changing this part:
```
--- /dev/null
+++ b/drivers/mtd/nand/raw/mt7621_nand.c
@@ -0,0 +1,1348 @@
```
which means this patch now becomes invalid.
You should change the 1348 to 1350 as two lines are added.

anton-ryzhov replied:
Sorry, I never worked with raw patches before. Thanks

981213 reviewed (commented):
@anton-ryzhov 
Correct. But...

981213 commented:
BTW could you run this patch through scripts/checkpatch.pl and see if it generates any warning about the curl braces?
```
./scripts/checkpatch.pl target/linux/ramips/patches-5.4/0300-mtd-rawnand-add-driver-support-for-MT7621-nand-flash.patch
```

anton-ryzhov commented:
No new warnings/errors since master version:
```
$ ./scripts/checkpatch.pl target/linux/ramips/patches-5.4/0300-mtd-rawnand-add-driver-support-for-MT7621-nand-flash.patch
WARNING: please write a paragraph that describes the config symbol fully
#27: FILE: drivers/mtd/nand/raw/Kconfig:394:
+config MTD_NAND_MT7621

WARNING: please write a paragraph that describes the config symbol fully
#35: FILE: drivers/mtd/nand/raw/Kconfig:402:
 config MTD_NAND_MTK

ERROR: do not use C99 // comments
#51: FILE: drivers/mtd/nand/raw/mt7621_nand.c:1:
+// SPDX-License-Identifier: GPL-2.0

WARNING: line over 80 characters
#738: FILE: drivers/mtd/nand/raw/mt7621_nand.c:688:
+					   const struct nand_data_interface *conf)

total: 1 errors, 3 warnings, 1371 lines checked

target/linux/ramips/patches-5.4/0300-mtd-rawnand-add-driver-support-for-MT7621-nand-flash.patch has style problems, please review.

If any of these errors are false positives, please report
them to the maintainer, see CHECKPATCH in MAINTAINERS.
```

LynxChaus commented:
There is one caller with NULL - mt7621_nfc_write_oob_hwecc() which intended to write OOB data, so we may bypass check mt7621_nfc_check_empty_page(nand, buf) when buf ptr is NULL.
Move check down to mt7621_nfc_write_page_hwecc() func.

981213 commented:
@LynxChaus write_oob_hwecc is supposed to write data and/or oob. The name is a bit confusing.
mt7621_nfc_check_empty_page skips a write with all 0xff content. We should still let check_empty_page to check oob if we want to keep the intended logic in write_oob_hwecc unchanged.

LynxChaus commented:
Read original report (link in first message). Called  mt7621_nfc_write_oob_hwecc() MUST write OOB data (bad block marker). So, checking OOB in mt7621_nfc_check_empty_page() is useless. It always true (buffer is emty (absent) and OOB data is present).

anton-ryzhov commented:
@LynxChaus 

How is that related to this PR at all? The `mt7621_nfc_check_empty_page` function may receive null-pointer but it access its data without pre-check. That's unsafe behaviour which causes a real bug.

You're assuming that this may happen only via `mt7621_nfc_write_oob_hwecc`, but why you are so sure? `mt7621_nfc_write_page_hwecc` is exported as `nand->ecc.write_oob`, who knows how it may be called. And it is irrelevant.

There may be a call to write OOB with all bytes `0xFF`, so `mt7621_nfc_check_empty_page` may return 0/1.

981213 commented:
@LynxChaus 
```
Called mt7621_nfc_write_oob_hwecc() MUST write OOB data (bad block marker). So, checking OOB in mt7621_nfc_check_empty_page() is useless. It always true (buffer is emty (absent) and OOB data is present).
```
If we just assume oob isn't empty if buf is null, we may also assume that when buf is present, the caller wants to write some data to the page. It seems to me that the entire check_empty_page routine becomes superfluous under this assumption, so I'm not sure whether this assumption is correct.
This check_empty_page itself is a hack. Either we keep the original intention unchanged just to fix the null pointer dereferencing (so that we don't accidentally break something else), or we should actually figure out what it's doing. Since this bug brick people's devices I'd prefer to get the former done to fix the bug first.

LynxChaus commented:
> This check_empty_page itself is a hack. 
So, remove it completely. There is two variants - kernel write buffer (data) and ECC engine MUST write ECC in OOB, or kernel write OOB data, so WHOLE page (with OOB) is erased+programmed and ECC engine MUST recalculate code and write it to OOB.

For what this hack exists?

anton-ryzhov commented:
@LynxChaus submit a separate PR?

981213 commented:
@LynxChaus 
> so WHOLE page (with OOB) is erased+programmed and ECC engine MUST recalculate code and write it to OOB.

Unfortunately NAND can't be erased per-page. An erase is performed on the whole block instead.
It seems that current kernel nand framework assumes that OOB isn't protected by ECC and can be written separately without being erased. jffs2 does this (writing the same page twice, one for data and the other for OOB) and a hack in mediatek sdk driver is to read the whole block(usually 64+ pages), erase the block, and then write every page back.
As we can't change hardware behavior to make it comply with existing software, a proper solution is to fix the nand framework so that it's aware it can't write data/OOB separatedly.

Anyway this seems to be another unrelated issue:

> For what this hack exists?

To be honest I don't know either. I'm calling it a hack only because it looks weird to me. The check_empty_page we are discussing here skips a write with all 0xff data/OOB. I don't know why such a call of write_oob would exist.

981213 commented:
Merged the fix into my staging tree at: https://git.openwrt.org/?p=openwrt/staging/981213.git
Thanks!

981213 commented:
@LynxChaus 
> For what this hack exists?

Got the answer from @hackpascal :
Linux kernel expects that when you write all 0xff into a page, the calculated ECC parity data should be 0xff as well. An example of such a case is writing data backed up from a different device with ECC enabled using mtd. Empty pages in that mtd copy will be 0xff. The software BCH code in linux achieved this by having a copy of precalculated ECC parity data of a full 0xff page, invert it, and then xor the actual ECC parity code with this precalculated value. (As a result, stored ECC parity data will be 0xff for a full 0xff page).
As mt7621 nand controller doesn't do that, this hack exists to replicate this behavior: When writing a full 0xff page, the page content should all be 0xff so that the nand page is left writable (empty).
